paddle/fluid/operators/conv_mkldnn_op.cc

```diff
@@ -216,72 +438,305 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
+                            framework::DataTypeTrait<int8_t>::DataType)
+                      : paddle::framework::ToMKLDNNDataType(
+                            framework::DataTypeTrait<uint8_t>::DataType);
+
+    if (mkldnninfo->force_fp32_output) {
+      dst_dt = paddle::framework::ToMKLDNNDataType(
+          framework::DataTypeTrait<float>::DataType);
+    }
+
+    if (mkldnninfo->fuse_residual_conn) {
+      auto residual = ctx.Input<Tensor>("ResidualData");
+      auto residual_dt = paddle::framework::ToMKLDNNDataType(residual->type());
+      if (dst_dt != residual_dt) dst_dt = residual_dt;
+    }
+    auto dst_md = platform::MKLDNNMemDesc(*convinfo->dst_tz, dst_dt,
+                                          chosen_memory_format);
+
+    // create a conv primitive descriptor and save it for usage in backward
+    if (bias) {
+      auto bias_md = platform::MKLDNNMemDesc(bias_tz, memory::data_type::s32,
+                                             memory::format::x);
+      mkldnninfo->conv_pd = ConvFwdPrimitiveDesc(
+          src_md, weights_md, bias_md, dst_md, *convinfo->strides,
+          *convinfo->paddings, *mkldnninfo->mkldnn_engine,
+          mkldnninfo->fuse_relu, mkldnninfo->fuse_residual_conn,
+          output_shift_scale, sum_scale, mkldnninfo->is_test);
+    } else {
+      mkldnninfo->conv_pd = ConvFwdPrimitiveDesc(
+          src_md, weights_md, dst_md, *convinfo->strides, *convinfo->paddings,
+          *mkldnninfo->mkldnn_engine, mkldnninfo->fuse_relu,
+          mkldnninfo->fuse_residual_conn, output_shift_scale, sum_scale,
+          mkldnninfo->is_test);
+    }
+    // Save conv_pd/src_memory/weights_memory for backward pass
+    dev_ctx->SetBlob(*mkldnninfo->key_conv_pd, mkldnninfo->conv_pd);
+
+    mkldnninfo->handler.reset(new platform::ConvMKLDNNHandler(
+        mkldnninfo->conv_pd, *dev_ctx, *mkldnninfo->mkldnn_engine,
+        *mkldnninfo->key));
+
+    // create mkldnn memory from input tensors (data/weights)
+    mkldnninfo->user_src_memory_p = mkldnninfo->handler->AcquireSrcMemory(
+        user_src_md, to_void_cast<T>(input_data));
+    auto user_weights_memory_p = mkldnninfo->handler->AcquireWeightsMemory(
+        user_weights_md, to_void_cast<K>(filter_data));
 
-      auto bias_memory_p =
-          handler.AcquireBiasMemoryFromPrimitive(user_bias_memory_p, pipeline);
-      conv_p = handler.AcquireConvolution(src_memory_p, weights_memory_p,
-                                          bias_memory_p, dst_memory_p);
+    // create reorder primitive if the input format is not the preferred one
+    mkldnninfo->src_memory_p =
+        mkldnninfo->handler->AcquireSrcMemoryFromPrimitive(
+            mkldnninfo->user_src_memory_p, *mkldnninfo->pipeline);
+
+    std::shared_ptr<mkldnn::memory> weights_memory_p;
+    int mask_reorder = is_multi_channel
+                           ? ((convinfo->g != 1) ? (1 << 1) + (1 << 0) : 1 << 0)
+                           : 0;
```

Comment:
Could we replace the bit shifts with the actual values? `3 : 1`

Reply:
It's according to MKL-DNN implementation for data format.